base_classes/nyaml/NXbeam.yaml

```diff
@@ -0,0 +1,258 @@
+doc: | 
+  Properties of the neutron or X-ray beam at a given location.
+  
+  This group is intended to be referenced
+  by beamline component groups within the :ref:`NXinstrument` group or by the :ref:`NXsample` group. This group is
+  especially valuable in storing the results of instrument simulations in which it is useful
+  to specify the beam profile, time distribution etc. at each beamline component. Otherwise,
+  its most likely use is in the :ref:`NXsample` group in which it defines the results of the neutron
+  scattering by the sample, e.g., energy transfer, polarizations. Finally, There are cases where the beam is
+  considered as a beamline component and this group may be defined as a subgroup directly inside
+  :ref:`NXinstrument`, in which case it is recommended that the position of the beam is specified by an
+  :ref:`NXtransformations` group, unless the beam is at the origin (which is the sample).
+  
+  Note that incident_wavelength and related fields can be a scalar values or arrays, depending on the use case.
+  To support these use cases, the explicit dimensionality of these fields is not specified, but it can be inferred
+  by the presense of and shape of accompanying fields, such as incident_wavelength_weights for a polychromatic beam.
+symbols: 
+    doc: |
+      These symbols coordinate datasets with the same shape.
+
+    nP: |
+      Number of scan points.
+
+    m: |
+      Number of channels in the incident beam spectrum, if known
+
+    c: |
+      Number of moments representing beam divergence (x, y, xy, etc.)
+
+category: base
+NXbeam:
+  distance(NX_FLOAT):
+    unit: NX_LENGTH
+    doc: |
+      Distance from sample. Note, it is recommended to use NXtransformations instead.
+  incident_energy(NX_FLOAT):
+    unit: NX_ENERGY
+    doc: |
+      Energy carried by each particle of the beam on entering the beamline component.
```

Comment:
The use of incident and final to me indicates we talk about in fact a beam that should be modelled as two instances

So instead of writing incident_energy and final_energy I would model NXbeam with a field energy
and then when NXbeam is used in an instance (e.g. an appdef) one would have incident(NXbeam) and final(NXbeam) this view is also substantiated by the statement "at a given location" because why is the beam upon entering the beamline component the same beam conceptually than the exiting beam, one could also argue that exiting beam is in fact a realization of the fact that there was in the first place a beam entering the beamline component and that beam was just modified upon interacting with the fields or stimuli applied by the beam component,
this would substantially reduce the amount of terms to have in the base class
the disadvantage is of course the refactoring needed in cases where so far one group beam kept all this but now it becomes two groups in an hdf5 file e.g. incipient and final but that is anyway assumed as this base class is to be used to describe a beam "at a location" so how stiff is this location argument (the entire beamline component is the "location" or the points where the beam enters or exits respectively.

The argument here is essentially borrowed from Adnam Hashibon; I tend to agree with him the situation that we feel there is a need to prefix the energy with "incipient" or "final" already indicates might be a good clue to guide us that we should talk and think about two instances (of the same concept?).

Reply:
The concept and strength of a glossary is that by compositing words we can construct and describe substantially more ideas and concepts with a language than compared as if we would just use the same 30k words in one language or glyphs with only one mapping of a concept per glyph or word. An electric_car is a special instance of car